### ISSUE TITLE
False positive for infinite_iter

### ISSUE THREAD

**Issue by michael-p** [rust-lang/rust-clippy#3538]
The following code causes clippy to trigger the `infinite_iter` lint but it's perfectly fine since `ArrayVec` only takes a limited amount (10 in this case) of items from the iterator (see [ArrayVec documentation](https://docs.rs/arrayvec/0.4.8/arrayvec/struct.ArrayVec.html#impl-FromIterator%3C%3CA%20as%20Array%3E%3A%3AItem%3E)
```rust
use arrayvec::ArrayVec;
use std::iter;

fn main() {
    let stuff: ArrayVec<[u8; 10]> = iter::repeat(22).collect();
    println!("{:?}", stuff); // -> [22, 22, 22, 22, 22, 22, 22, 22, 22, 22]
}
```
Note that `infinite_iter` is set to `deny` by default, causing compilation to abort with an error.

Clippy version: `clippy 0.0.212 (2e26fdc2 2018-11-22)`

**Comment by oli-obk** [comment]
We should limit the lint to

* `Vec`
* `VecDeque`
* `BTreeSet`, `BTreeMap`
* `HashSet`, `HashMap`

which are known to be infinitely allocating on infinite iterators.

**Comment by michael-p** [comment]
Sounds good to me! Should probably also include `LinkedList` and `BinaryHeap`, for the sake of completeness.
